core/src/main/java/com/linecorp/armeria/common/CompositeStringMultimap.java

```diff
@@ -0,0 +1,1254 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import static com.linecorp.armeria.common.StringMultimap.HASH_CODE_SEED;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterators;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.handler.codec.DateFormatter;
+
+/**
+ * The composite container implementation which wraps {@link StringMultimap}s
+ * to avoid expensive copy operations.
+ *
+ * <p>Adds entries only on {@code appender} and marks entries as removed only on {@code removed}
+ * without modifying wrapped {@link StringMultimap}s to avoid expensive copy operations.
+ *
+ * <p>So any further modifications to this don't break the immutability of wrapped {@link StringMultimap}s.
+ *
+ * @param <IN_NAME> the type of the user-specified names, which may be more permissive than {@link NAME}
+ * @param <NAME> the actual type of the names
+ *
+ * @see CompositeHttpHeadersBase
+ */
+abstract class CompositeStringMultimap<IN_NAME extends CharSequence, NAME extends IN_NAME>
```

Comment:
I think is should be `Multi'M'ap`.

Reply:
<img width="204" alt="image" src="https://github.com/line/armeria/assets/34854527/4ea46f45-b8a6-4d1b-9c4d-379e6fd8c647">

Oh I just copy&paste it from `StringMultimap`. should we use `Multi'M'ap` here?